### PR TITLE
Create a function for calculating CIIP Incentive Payments per product

### PR DESCRIPTION
**Pull request by hamzajaved**

### CODE REVIEW

File: schema/deploy/ciip_functions/ciip_incentive_payment.sql [hunk 1]
```diff
@@ -0,0 +1,112 @@
+-- Deploy ggircs-portal:ciip_functions/ciip_incentive_payment to pg
+-- requires: tables/benchmark
+-- requires: views/ciip_production
+-- requires: views/ciip_electricity_and_heat
+
+begin;
+
+-- Function takes application id and version number:
+create or replace function ggircs_portal.ciip_incentive_payment(app_id numeric , version_no numeric )
```
Comment (matthieu-foucault) [inline schema/deploy/ciip_functions/ciip_incentive_payment.sql]:
```suggestion
create or replace function ggircs_portal.ciip_incentive_payment(app_id int , version_no int )
```

File: schema/deploy/ciip_functions/ciip_incentive_payment.sql [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  begin
+     -- Define placeholder variables
+     em_alloc_elec = 0;
+     em_alloc_heat = 0;
+     incentive_ratio_min = 0;
+     incentive_ratio_max = 1;
+
+     -- Get emissions for facility
+     select sum(annual_co2e) into em_facility from ggircs_portal.ciip_emission
+     where version_number = version_no and application_id = app_id;
```
Comment (matthieu-foucault) [inline schema/deploy/ciip_functions/ciip_incentive_payment.sql]:
```suggestion
     where version_number = version_no and application_id = app_id and gas_type != 'CO2';
```
biogenic emissions should not be included

File: schema/deploy/ciip_functions/ciip_incentive_payment.sql [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+     -- Loop over products
+     for product in select * from ggircs_portal.ciip_production
+                    where version_number = version_no and application_id = app_id
+     loop
+
+        -- Calculate Emissions for Product (EmProd)
+        -- EmProd = EmFacility * EmAlloc(P,F) + EmImportedElec * EmAlloc(IE) + EmImportedHeat * EmAlloc(IH)
+        em_product = (em_facility * product.production_allocation_factor/100)
+                   + (em_electricity * em_alloc_elec)
+                   + (em_heat * em_alloc_elec);
```
Comment (matthieu-foucault) [inline schema/deploy/ciip_functions/ciip_incentive_payment.sql]: This is where we'll need an "if benchmark includes energy import"

File: schema/deploy/ciip_functions/ciip_incentive_payment.sql [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+        -- Calculate Emission Intensity
+        em_intensity = em_product / product.quantity;
+
+        -- Get Incentive Ratio Max and Min, BM and ET for product from Benchmark table
+        select * into benchmark_data from ggircs_portal.benchmark
+        where product_id = product.product_id;
```
Comment (matthieu-foucault) [inline schema/deploy/ciip_functions/ciip_incentive_payment.sql]: This needs to be filtered based on the application reporting year (has to be between start_reporting_year and end_reporting_year, inclusive)

File: schema/deploy/ciip_functions/ciip_incentive_payment.sql (hunk 3 above)
Comment (hamzajaved) [reply schema/deploy/ciip_functions/ciip_incentive_payment.sql]: as discussed, can these allocation values be zero if there is no energy import?

Comment (hamzajaved) [reply schema/deploy/ciip_functions/ciip_incentive_payment.sql]: @matthieu-foucault

Comment (matthieu-foucault) [reply schema/deploy/ciip_functions/ciip_incentive_payment.sql]: I think so

File: schema/deploy/views/ciip_carbon_tax_calculation.sql [hunk 5]
```diff
@@ -71,8 +71,6 @@ begin;
     from carbon_tax_variables
  );
 
-grant all on table ggircs_portal.ciip_carbon_tax_calculation to ciip_administrator, ciip_analyst, ciip_industry_user, ciip_guest;
```
Comment (matthieu-foucault) [inline schema/deploy/views/ciip_carbon_tax_calculation.sql]: was this removed on purpose? If so, why isn't it needed?

Comment (hamzajaved) [reply schema/deploy/views/ciip_carbon_tax_calculation.sql]: Hmm I didn't explicitly remove it. I pulled some files from develop, that may have removed it.

**Pull request merged**